### PR TITLE
[RISCV] Let -data-sections also work on sbss/sdata sections

### PR DESCRIPTION
**Pull request by KaiYG**
Add an unique suffix to .sbss/.sdata if -fdata-sections.
Without assigning an unique .sbss/.sdata section to each symbols, a linker may not be able to remove unused part when gc-section since all used and unused symbols are all mixed in the same .sbss/.sdata section.
I believe this also matches the behavior of gcc.

### CODE REVIEW

**Comment by KaiYG** [comment]
I've created [an upstream PR](https://github.com/llvm/llvm-project/pull/87040) as well, so I will wait a bit longer to see if I can cherry-pick it from upstream instead.

**Comment by KaiYG** [comment]
Ping.
I think this patch will save code size on both bfd/lld linker, and a similar patch has been done on the Hexagon backend.

**Review by kito-cheng — APPROVED** [review]
LGTM :)

**Comment by jonathonpenix** [comment]
Are there any outstanding concerns on this patch? We recently saw this issue and I was wondering if it would be ok to merge this at this point. Thanks!

**Comment by lenary** [comment]
@topperc yeah, I was thinking that too - I'll take a closer look.

**Review by MaskRay — CHANGES_REQUESTED** [review]
Before we do more work with sdata/sbss I think we should consider 
#83093

or at least ensure regular non-baremetal targets don't get weird non-zero -msmall-data-limit=

File: llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp [hunk 1]
```diff
@@ -104,10 +104,36 @@ bool RISCVELFTargetObjectFile::isGlobalInSmallSection(
...
-    return SmallDataSection;
+  if (isGlobalInSmallSection(GO, TM)) {
+    // Emit a unique section for sdata/sbss when -fdata-section.
+    bool EmitUniquedSection = TM.getDataSections();
+
+    if (Kind.isBSS()) {
+      if (EmitUniquedSection) {
+        StringRef Prefix(".sbss");
+        SmallString<128> Name(Prefix);
+        Name.append(".");
```
Comment (petrhosek) [inline llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp:115]:
```suggestion
        SmallString<128> Name(".sbss.");
```

File: llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp [hunk 2]
```diff
@@ -104,10 +104,36 @@ bool RISCVELFTargetObjectFile::isGlobalInSmallSection(
...
+      }
+
+      return SmallBSSSection;
+    }
+
+    if (Kind.isData()) {
+      if (EmitUniquedSection) {
+        StringRef Prefix(".sdata");
+        SmallString<128> Name(Prefix);
+        Name.append(".");
```
Comment (petrhosek) [inline llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp:128]:
```suggestion
        SmallString<128> Name(".sdata.");
```

**Comment by petrhosek** [comment]
There's a change in semantics which I think is undesirable due to https://github.com/llvm/llvm-project/pull/87040/files#diff-16406b22032eaeb81527fc5cf1938dd0981c926fcad2f5b2e9a9f9db7d31f03cR79-R82

For example, if you have the following:
```
__attribute__((section(".bss"))) int foo = 0;
```
`foo` will end up in `.bss` even when `-fdata-sections` is set.

However, if you have the following:
```
__attribute__((section(".sbss"))) int foo = 0;
```
`foo` will end up in `.sbss.foo` after this change and I think that's undesirable.

If a symbol has an explicit section, we should honor that. This could be achieved simply by always returning `false` if https://github.com/llvm/llvm-project/pull/87040/files#diff-16406b22032eaeb81527fc5cf1938dd0981c926fcad2f5b2e9a9f9db7d31f03cL76 is true.

**Comment by KaiYG** [comment]
> There's a change in semantics which I think is undesirable due to https://github.com/llvm/llvm-project/pull/87040/files#diff-16406b22032eaeb81527fc5cf1938dd0981c926fcad2f5b2e9a9f9db7d31f03cR79-R82

Good catch. But I patched this issue in another way, as I think `isGlobalInSmallSection` should still return `true` for explicit sdata/sbss sections, and `EmitUniquedSection` should be set to `false` instead.

File: llvm/test/CodeGen/RISCV/sdata-sections.ll [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+; RUN: llc -mtriple=riscv64 -data-sections < %s | FileCheck -check-prefix=RV64 %s
+
+; Append an unique name to each sdata/sbss section when -data-section.
+
+@v = dso_local global i32 0, align 4
+@r = dso_local global i64 7, align 8
+
+; If a symbol has an explicit sdata/sbss section name, we should honor it.
+@vv = dso_local global i32 0, section ".sbss", align 4
+@rr = dso_local global i64 7, section ".sdata", align 8
```
Comment (lenary) [inline llvm/test/CodeGen/RISCV/sdata-sections.ll:11]: please also test with custom names that aren't `.sbss` or `.sdata`.

File: llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp [hunk 4]
```diff
@@ -104,10 +104,34 @@ bool RISCVELFTargetObjectFile::isGlobalInSmallSection(
...
   // Handle Small Section classification here.
-  if (Kind.isBSS() && isGlobalInSmallSection(GO, TM))
-    return SmallBSSSection;
-  if (Kind.isData() && isGlobalInSmallSection(GO, TM))
-    return SmallDataSection;
+  if (isGlobalInSmallSection(GO, TM)) {
+    // Emit to an unique sdata/sbss section when -fdata-section is set.
+    // However, if a symbol has an explicit sdata/sbss section, place it in that
+    // section.
+    bool EmitUniquedSection = TM.getDataSections() && !GO->hasSection();
```
Comment (lenary) [inline llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp:111]: I think this is still wrong, and that the `!GO->hasSection()` check needs to take priority over the `isGlobalInSmallSection` - I noted in the test where I think things will still go wrong.

**Review by lenary — COMMENTED** [review]
I re-looked at this code and there are still issues with globals with explicit sections

File: llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp (hunk 4 above)
Comment (KaiYG) [reply llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp:111]: I think `isGlobalInSmallSection` always return `false` unless the explicit section is a `.sdata` or `.sbss`, so I meant to let symbols with `__attribute__((section(".sdata")))`/`__attribute__((section(".sbss")))` neglect -fdata-sections and directly use `SmallDataSection`/`SmallBSSSection`.
Please correct me if I misunderstand, thanks.

Comment (lenary) [reply llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp:111]: Ah, I see what's going on.

**Review by lenary — APPROVED** [review]
LGTM

**Review by petrhosek — APPROVED** [review]

**Review by MaskRay — APPROVED** [review]

**Comment by lenary** [comment]
@KaiYG right now, if i try to squash merge, the commit will use your `@users.noreply.github.com` email address, which LLVM doesn't want in the commit history - please can you follow the instructions here: [Setting your commit email address on GitHub](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address#setting-your-commit-email-address-on-github) to change it to your andestech email? Then I will squash/merge.

**Comment by lenary** [comment]
Thanks, that did the right thing :)

**Comment by lenary** [comment]
I forgot that a question about backporting this change came up in the RISC-V sync-up. I think we're too late to backport this to 19.x, as it isn't a bugfix, it's an improvement.

**Pull request merged**